storage/gcsupload/gcsupload.go

```diff
@@ -0,0 +1,156 @@
+//	    the path to the source
+//
+// For example:
+//  gcsupload --project-id gcs-samples --source ~/Desktop/birthdayPic.jpg --bucket gcs-cli-test
+//  URL: https://storage.googleapis.com/gcs-cli-test/birthdayPic.jpg
+//  Size: 865096
+//  MD5: 5b6c7b4aed837e8ed0f9950564a10b32
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+
+	"cloud.google.com/go/storage"
+
+	"golang.org/x/net/context"
+)
+
+func main() {
+	// Prevent log from printing out time information
+	log.SetFlags(0)
+
+	var projectID, bucket, source, name string
+	var public bool
+
+	flag.StringVar(&bucket, "bucket", "", "the bucket to upload content to")
+	flag.StringVar(&projectID, "project-id", "", "the ID of the GCP project to use")
+	flag.StringVar(&source, "source", "", "the path to the source")
+	flag.StringVar(&name, "name", "", "the name of the file to be stored on GCS")
+	flag.BoolVar(&public, "public", true, "whether the item should be available publicly")
+	flag.Parse()
+
+	// If they haven't set the bucket or projectID nor specified
+	// in the environment, then fail if missing.
+	bucket = envOrFatal("GOLANG_SAMPLES_BUCKET", bucket)
+	projectID = envOrFatal("GOLANG_SAMPLES_PROJECT_ID", projectID)
+
+	var r io.Reader
+	if source == "" {
+		r = os.Stdin
+		log.Printf("Reading from stdin...")
+	} else {
+		f, err := os.Open(source)
+		if err != nil {
+			log.Fatal(err)
+		}
+		defer f.Close()
+		r = f
+	}
+
+	if name == "" {
+		if source != "" {
+			name = filepath.Base(source)
+		} else {
+			name = "test-sample"
+		}
+	}
+
+	ctx := context.Background()
+	_, objAttrs, err := upload(ctx, r, projectID, bucket, name, true)
+	if err != nil {
+		log.Fatalf("Upload err: %v", err)
+	}
+	log.Printf("URL: %s\nSize: %d\nMD5: %x\nobjAttrs: %+v\n", objectURL(objAttrs), objAttrs.Size, objAttrs.MD5, objAttrs)
+}
+
+func objectURL(objAttrs *storage.ObjectAttrs) string {
+	return fmt.Sprintf("https://storage.googleapis.com/%s/%s", objAttrs.Bucket, objAttrs.Name)
+}
+
+func upload(ctx context.Context, r io.Reader, projectID, bucket, name string, public bool) (*storage.ObjectHandle, *storage.ObjectAttrs, error) {
+	bh, _, err := ensureBucketExists(ctx, projectID, bucket)
+	if err != nil {
+		return nil, nil, err
+	}
+	obj := bh.Object(name)
+	w := obj.NewWriter(ctx)
+	if _, err := io.Copy(w, r); err != nil {
+		return nil, nil, err
+	}
+	if err := w.Close(); err != nil {
+		return nil, nil, err
+	}
+
+	if public {
+		if err := obj.ACL().Set(ctx, storage.AllUsers, storage.RoleReader); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	attrs, err := obj.Attrs(ctx)
+	return obj, attrs, err
+}
+
+// ensureBucketExists first checks if a bucket exists on GCS. If that check
+// fails, it then attempts to make the bucket, and sets the default ACL for objects
+// to either "private" or "public" depending on if objectsPublicByDefault is set.
+func ensureBucketExists(ctx context.Context, projectID, bucket string) (bh *storage.BucketHandle, existed bool, err error) {
```

Comment:
is this sample showing how to create buckets or upload files?

I'd just handle this error and tell people to create the bucket, rather than conflating this program to both creating buckets and uploading objects.